### PR TITLE
feat: manager-agnostic changelog

### LINKED ISSUE #2101: Refactor Changelog to be manager-independent
Currently the changelog approach is manager-specific - we have one for npm and one for pypi. Instead, we should make it manager-agnostic so that `changelog.md` files or GitHub releases can be used/included  for any package manager that reports the correct source.

Proposal:
- Changelog should be manager-agnostic, and work anytime the datasource includes a valid source repository with an upgrade
- Datasource should include an ordered list of "intermediate versions" with lookup, which are passed through and used by changelog function. That way changelog doesn't need to duplicate functionality by looking up the list of releases to document. I'm unsure if intermediate releases should include both the fromVersion and toVersion too, or only the ones in-between. It doesn't really matter so should be whatever is most convenient to changelog
- We currently use "repositoryUrl" for this but this could be renamed, e.g. to "source" if we think there will be non-repository source URLs in future

Logic flow is:
- datasource/lookupUpdates will return an additional field - intermediateVersions (or some other name) with dependency lookup
- datasource will also return `repositoryUrl` - or some other name - representing the source to get release notes from. Right now this can just be github-only
- changelog will look if `repositoryUrl` is present and if so then it will look for changelog files and/or releases

### PR DESCRIPTION
**Pull request by kayoub5**
Closes #2101 #2092

### CODE REVIEW

File: lib/datasource/github.js [hunk 1]
```diff
@@ -47,16 +44,14 @@ async function getDependency(purl) {
...
-  }
   const dependency = {
     repositoryUrl: 'https://github.com/' + repo,
-    versions: {},
   };
-  versions.forEach(version => {
-    dependency.versions[version] = {};
-  });
+  // TODO: should Github receive the versionScheme as input?
+  // versions should be already sorted by GIT
```
Comment (rarkins) [inline lib/datasource/github.js]: I don't believe that can be true. Git would need to be node semver-aware for that to be true. e.g. the Node.js github repo released 10.x a while ago yet 8.x still gets updates. I don't see how git would sort correctly using either (a) chronological order, or (b) alphabetical order.

I am OK with passing `versionScheme` to datasource and having it sort and/or filter based on that.

File: lib/datasource/github.js [hunk 2]
```diff
@@ -47,16 +44,14 @@ async function getDependency(purl) {
...
+  dependency.releases = versions.map(version => ({
+    version: version.replace(/^v(\d)/, '$1'),
```
Comment (rarkins) [inline lib/datasource/github.js]: This datasource shouldn't filter `v` prefix *every time*. There may be package managers using the github datasource that depend on having the exact tag, and not a "cleaned" version of it.

File: lib/datasource/npm.js [hunk 3]
```diff
@@ -185,17 +188,22 @@ async function getDependencyInner(name, retries = 5) {
...
       latestVersion: res['dist-tags'].latest,
       repositoryUrl,
-      versions: res.versions,
-      'dist-tags': res['dist-tags'],
       'renovate-config': latestVersion['renovate-config'],
     };
-    Object.keys(dep.versions).forEach(version => {
-      const v = {};
+    const versions = Object.keys(res.versions)
+      .filter(version => isVersion(version))
```
Comment (rarkins) [inline lib/datasource/npm.js]: This is fine to leave in, but do you think npmjs ever returns a non-semver version?

File: lib/datasource/github.js [hunk 4]
```diff
@@ -47,16 +44,14 @@ async function getDependency(purl) {
...
   const dependency = {
     repositoryUrl: 'https://github.com/' + repo,
-    versions: {},
   };
-  versions.forEach(version => {
-    dependency.versions[version] = {};
-  });
+  // TODO: should Github receive the versionScheme as input?
+  // versions should be already sorted by GIT
+  dependency.releases = versions.map(version => ({
```
Comment (rarkins) [inline lib/datasource/github.js]: I assume it's intentional to make releases now an array containing `version` as a field rather than an object with `version` as the key? Maybe we can separate that into its own commit/PR first and reduce the number of files that this PR changes.

File: lib/datasource/pypi.js [hunk 5]
```diff
@@ -23,13 +26,15 @@ async function getDependency(purl) {
...
-      Object.keys(dep.releases).forEach(release => {
-        dependency.versions[release] = {
-          date: (dep.releases[release][0] || {}).upload_time,
-        };
-      });
+      const versions = Object.keys(dep.releases)
+        .filter(version => isVersion(version))
+        .sort(sortVersions);
+      dependency.releases = versions.map(version => ({
+        version: isVersion(version),
```
Comment (rarkins) [inline lib/datasource/pypi.js]: Should we remove the requirement that `isVersion` returns a clean version and instead make it more obvious with `sanitize()` function for each versioning scheme?

File: lib/workers/pr/changelog/source-github.js [hunk 6]
```diff
@@ -42,12 +42,13 @@ async function getRepositoryHead(repository, version) {
   if (version.gitHead) {
     return version.gitHead;
   }
-  if (!version.date) {
+  const date = version.date || version.time;
```
Comment (rarkins) [inline lib/workers/pr/changelog/source-github.js]: Is this caused by inconsistency in our datasource code? Should we standardise this per-datasource now or later?

**Review by rarkins — CHANGES_REQUESTED** [review]

**Comment by rarkins** [comment]
@kayoub5 I have merged most changes piece by piece into master so that this PR is reduced to its main change. However one thing I did was different to how you had it - instead of including *every* release - which could include hundreds of irrelevant old ones, irrelevant pre-releases, etc - I instead just included a subset that match the `filteredVersions` during lookup. The issue with this is that your current code expects that the "fromVersion" release is also present, and checks for a length of at least 2 releases, whereas now the length may be as short as 1. Are you able to rewrite the logic to accommodate or do we need to add the fromVersion's release object too?

**Comment by rarkins** [comment]
Ignore the above comment - I have refactored it to return releases for all versions from `fromVersion` onwards

**Comment by rarkins** [comment]
Looks like I may have broken the `gitHead`/`gitRef` functionality. A good package to use to test this is `coveralls`, as they only have Releases for a few versions and nothing for the rest.

**Review by rarkins — APPROVED** [review]

**Comment by kayoub5** [comment]
@rarkins fixed the bugs, you may review the changes again
